File: flutter/lib/src/native/platform/send_encoding_parameters.dart
import 'dart:ffi';

import 'package:ffi/ffi.dart';
import 'package:medea_flutter_webrtc/medea_flutter_webrtc.dart';

import 'package:medea_jason/src/native/ffi/foreign_value.dart';
import 'package:medea_jason/src/native/ffi/native_string.dart';
import 'send_encoding_parameters.g.dart' as bridge;

/// Registers [SendEncodingParameters] related functions in Rust.
void registerFunctions(DynamicLibrary dl) {
  bridge.registerFunction(
    dl,
    newSendEncodingParameters: _newSendEncodingParameters,
    getRid: _getRid,
    setActive: _setActive,
    getActive: _getActive,
    setMaxBitrate: _setMaxBitrate,
    getMaxBitrate: _getMaxBitrate,
    setScaleResolutionDownBy: _setScaleResolutionDownBy,
    getScaleResolutionDownBy: _getScaleResolutionDownBy,
    setScalabilityMode: _setScalabilityMode,
    getScalabilityMode: _getScalabilityMode,
  );
}

/// Creates new [SendEncodingParameters].
SendEncodingParameters _newSendEncodingParameters(
  Pointer<Utf8> rid,
  bool active,
) {
  return SendEncodingParameters.create(rid.nativeStringToDartString(), active);
}

/// Returns [SendEncodingParameters.rid] from the provided
/// [SendEncodingParameters].
Pointer<Utf8> _getRid(Object encoding) {
  encoding as SendEncodingParameters;
  return encoding.rid.toNativeUtf8();
}

/// Sets [SendEncodingParameters.active] of the provided
/// [SendEncodingParameters].
void _setActive(Object encoding, bool active) {
  encoding as SendEncodingParameters;
  encoding.active = active;
}

/// Returns [SendEncodingParameters.active] of the provided
/// [SendEncodingParameters].
bool _getActive(Object encoding) {
  encoding as SendEncodingParameters;
  return encoding.active;
}

/// Sets [SendEncodingParameters.maxBitrate] of the provided
/// [SendEncodingParameters].
void _setMaxBitrate(Object encoding, int maxBitrate) {
  encoding as SendEncodingParameters;
  encoding.maxBitrate = maxBitrate;
}

/// Returns [SendEncodingParameters.maxBitrate] of the provided
/// [SendEncodingParameters].
Pointer _getMaxBitrate(Object encoding) {
  encoding as SendEncodingParameters;

  if (encoding.maxBitrate != null) {
    return ForeignValue.fromInt(encoding.maxBitrate!).intoRustOwned();
  } else {
    return ForeignValue.none().intoRustOwned();
  }
}

/// Sets [SendEncodingParameters.scaleResolutionDownBy] of the provided
/// [SendEncodingParameters].
void _setScaleResolutionDownBy(Object encoding, double scaleResolutionDownBy) {
  encoding as SendEncodingParameters;
  encoding.scaleResolutionDownBy = scaleResolutionDownBy;
}

/// Returns [SendEncodingParameters.scaleResolutionDownBy] of the provided
/// [SendEncodingParameters].
double _getScaleResolutionDownBy(Object encoding) {
  encoding as SendEncodingParameters;

  if (encoding.scaleResolutionDownBy != null) {
    return encoding.scaleResolutionDownBy!;
  } else {
    return 1.0;
  }
}

/// Sets [SendEncodingParameters.scalabilityMode] of the provided
/// [SendEncodingParameters].
void _setScalabilityMode(Object encoding, Pointer<Utf8> scalabilityMode) {
  encoding as SendEncodingParameters;
  encoding.scalabilityMode = scalabilityMode.nativeStringToDartString();
}

/// Returns [SendEncodingParameters.scalabilityMode] of the provided
/// [SendEncodingParameters].
Pointer<Utf8> _getScalabilityMode(Object encoding) {
  encoding as SendEncodingParameters;

  if (encoding.scalabilityMode != null) {
    return encoding.scalabilityMode!.toNativeUtf8();
  } else {
    return ''.toNativeUtf8();
  }
}

File: src/platform/wasm/send_encoding_parameters.rs
//! Wrapper around [RTCRtpEncodingParameters][0].
//!
//! [0]: https://w3.org/TR/webrtc#dom-rtcrtpencodingparameters

use derive_more::{From, Into};
use web_sys::RtcRtpEncodingParameters;

/// Wrapper around [RTCRtpEncodingParameters][0] providing handy methods for its
/// direction changes.
///
/// [0]: https://w3.org/TR/webrtc#dom-rtcrtpencodingparameters
#[derive(Clone, Debug, From, Into)]
pub struct SendEncodingParameters(RtcRtpEncodingParameters);

impl SendEncodingParameters {
    /// Creates new [`SendEncodingParameters`].
    #[expect(clippy::needless_pass_by_value, reason = "`cfg` code uniformity")]
    #[must_use]
    pub fn new(rid: String, active: bool) -> Self {
        let params = RtcRtpEncodingParameters::new();
        params.set_rid(&rid);
        params.set_active(active);
        Self(params)
    }

    /// Returns the underlying [`RtcRtpEncodingParameters`] of these
    /// [`SendEncodingParameters`].
    #[must_use]
    pub const fn handle(&self) -> &RtcRtpEncodingParameters {
        &self.0
    }

    /// Returns [RID] of these [`SendEncodingParameters`].
    ///
    /// [RID]: https://w3.org/TR/webrtc#dom-rtcrtpcodingparameters-rid
    #[must_use]
    pub fn rid(&self) -> Option<String> {
        self.0.get_rid()
    }

    /// Sets [activeness][1] of these [`SendEncodingParameters`].
    ///
    /// [1]: https://w3.org/TR/webrtc#dom-rtcrtpencodingparameters-active
    pub fn set_active(&self, active: bool) {
        self.0.set_active(active);
    }

    /// Returns [activeness][1] of these [`SendEncodingParameters`].
    ///
    /// [1]: https://w3.org/TR/webrtc#dom-rtcrtpencodingparameters-active
    #[must_use]
    pub fn active(&self) -> bool {
        // default is true according to spec
        self.0.get_active().unwrap_or(true)
    }

    /// Sets [maxBitrate][1] of these [`SendEncodingParameters`].
    ///
    /// [1]: https://w3.org/TR/webrtc#dom-rtcrtpencodingparameters-maxbitrate
    pub fn set_max_bitrate(&self, max_bitrate: u32) {
        self.0.set_max_bitrate(max_bitrate);
    }

    /// Returns [maxBitrate][1] of these [`SendEncodingParameters`].
    ///
    /// [1]: https://w3.org/TR/webrtc#dom-rtcrtpencodingparameters-maxbitrate
    #[must_use]
    pub fn max_bitrate(&self) -> Option<u32> {
        self.0.get_max_bitrate()
    }

    /// Sets [scaleResolutionDownBy][1] of these [`SendEncodingParameters`].
    ///
    /// [1]: https://tinyurl.com/ypzzc75t
    pub fn set_scale_resolution_down_by(&self, scale_resolution_down_by: f32) {
        self.0.set_scale_resolution_down_by(scale_resolution_down_by);
    }

    /// Returns [scaleResolutionDownBy][1] of these [`SendEncodingParameters`].
    ///
    /// [1]: https://tinyurl.com/ypzzc75t
    pub fn scale_resolution_down_by(&self) -> f64 {
        self.0.get_scale_resolution_down_by().map_or(1.0, Into::into)
    }

    /// Sets [scalabilityMode][1] of these [`SendEncodingParameters`].
    ///
    /// [1]: https://tinyurl.com/3zuaee45
    #[expect(clippy::needless_pass_by_value, reason = "`cfg` code uniformity")]
    pub fn set_scalability_mode(&self, scalability_mode: String) {
        self.0.set_scalability_mode(&scalability_mode);
    }

    /// Returns [scalabilityMode][1] of these [`SendEncodingParameters`].
    ///
    /// [1]: https://tinyurl.com/3zuaee45
    #[must_use]
    pub fn scalability_mode(&self) -> Option<String> {
        self.0.get_scalability_mode()
    }
}
